tests/contrib/algoliasearch/test.py

```diff
@@ -0,0 +1,117 @@
+import algoliasearch
+import algoliasearch.index as index_module
+from ddtrace import config
+from ddtrace.contrib.algoliasearch.patch import (SEARCH_SPAN_TYPE, patch,
+                                                 unpatch)
+from ddtrace.pin import Pin
+from tests.base import BaseTracerTestCase
+
+
+class AlgoliasearchTest(BaseTracerTestCase):
+    def setUp(self):
+        super(AlgoliasearchTest, self).setUp()
+
+        # dummy values
+        def search(self, query, args=None, request_parameters=None):
+            return {
+                'hits': [
+                    {
+                        'dummy': 'dummy'
+                    }
+                ],
+                'processingTimeMS': 23,
+                'nbHits': 1,
+                'hitsPerPage': 20,
+                'exhaustiveNbHits': True,
+                'params': 'query=xxx',
+                'nbPages': 1,
+                'query': 'xxx',
+                'page': 0
+            }
+
+        # Algolia search is a non free SaaS application, it isn't possible to add it to the
+        # docker environment to enable a full-fledged integration test. The next best option
+        # is to mock out the search method to prevent it from making server requests
+        index_module.Index.search = search
+        client = algoliasearch.algoliasearch.Client('X', 'X')
+        index = client.init_index('test_index')
+        patch()
+        Pin.override(index, tracer=self.tracer)
+
+        # use this index only to properly test stuff
+        self.index = index
+
+    def tearDown(self):
+        super(AlgoliasearchTest, self).tearDown()
+        unpatch()
+
+    def reset(self):
+        super(AlgoliasearchTest, self).reset()
+
+    def test_algoliasearch(self):
+        self.index.search(
+            'test search',
+            args={'attributesToRetrieve': 'firstname,lastname', 'unsupportedTotallyNewArgument': 'ignore'}
+        )
+
+        spans = self.get_spans()
+        self.reset()
+
+        assert len(spans) == 1
```

Comment:
[self.assert_structure()](https://github.com/DataDog/dd-trace-py/blob/4313f388383b90ccf2bcbca9d7ef1c400c827ece/tests/utils/span.py#L367) is pretty cool.

```python
self.index.search(....)

self.assert_span_count(1)
self.assert_structure(dict(
    service='algoliasearch',
    name='algoliasearch.search',
    span_type=SEARCH_SPAN_TYPE,
    error=0,
    meta={
        'query_args.attributes_to_retrieve': 'firstname,lastname',
    },
    metrics={
        'processing_time_ms': 23,
        'number_of_hits': 1,
    },
))
```

To ensure extra meta doesn't end up in `meta`/`metrics`, you can do:

```python
self.assert_structure('same as above but without meta or metrics')
root = self.get_root_span()

# Assert these and only these attributes exist in `meta`
root.assert_meta({
    'query_args.attributes_to_retrieve': 'firstname,lastname',
}, exact=True)
root.assert_metrics({

}, exact=True)
```

These are all fairly new patterns that haven't been adopted by every test suite yet.

Reply:
btw, the above is a suggestion, changing doesn't change anything that you are asserting here, everything here is fine, just pointing out the new patterns we have.

Reply:
I think we have an overabundance of pattern silos, I'm hesitant to introduce more. I think one thing the code could benefit from is some enforcement of patterns; if new patterns are to be introduced they should be accompanied with a refactoring effort instead of being dropped in new places.

Ultimately, this is your call, so let me know how you'd like me to proceed.

Reply:
Yeah, that was the hopeful idea to bring it across everything, but we didn't have the time to invest in that at the time, so we are trying to update when we can... not the best approach but what we did :/